clients/drpoints/drpoints.cpp

```diff
@@ -0,0 +1,439 @@
+    // Insert call to instrumentation function that saves the current BBV.
+    dr_insert_clean_call(drcontext, bb, inst, (void *)save_bbv, /*save_fpstate=*/false,
+                         0);
+    MINSERT(bb, inst, skip_call);
+    if (drreg_unreserve_aflags(drcontext, bb, inst) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+    if (scratch != DR_REG_NULL) {
+        if (drreg_unreserve_register(drcontext, bb, inst, scratch) != DRREG_SUCCESS)
+            DR_ASSERT(false);
+    }
+#    else
+#        error NYI
+#    endif
+#else
+    // Default to a clean call to the instrumentation function that increments the
+    // counter, checks if the user defined instruction interval is reached, and if so
+    // saves the current BBV.
+    dr_insert_clean_call(drcontext, bb, inst, (void *)increment_counters_and_save_bbv,
+                         /*save_fpstate=*/false, 2, OPND_CREATE_INTPTR(bb_id),
+                         OPND_CREATE_INTPTR(bb_size));
+#endif
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+event_exit(void)
+{
+    bool generate_file = !no_out_bbv_file.get_value();
+    file_t bbvs_file = INVALID_FILE;
+    if (generate_file) {
+        // Get the current working directory where drrun is executing. We save the .bbv
+        // file there.
+        char cwd[MAXIMUM_PATH];
+        if (!dr_get_current_directory(cwd, sizeof(cwd)))
+            FATAL("ERROR: dr_get_current_directory() failed");
+
+        // Create and open the drpoints.PROC_BIN_NAME.PID.UNIQUE_ID.bbv file.
+        char buf[MAXIMUM_PATH];
+        bbvs_file =
+            drx_open_unique_appid_file(cwd, dr_get_process_id(), "drpoints", "bbv",
+                                       DR_FILE_WRITE_REQUIRE_NEW, buf, sizeof(buf));
+        if (bbvs_file == INVALID_FILE)
+            FATAL("ERROR: unable to create BBVs file");
+    }
+
+    // Define the format strings we use to write the file.
+    const char *one_pair_only = "T:%d:%d \n";
+    const char *first_pair = "T:%d:%d ";
+    const char *middle_pair = ":%d:%d ";
+    const char *last_pair = ":%d:%d \n";
+
+    for (uint i = 0; i < bbvs.entries; ++i) {
+        drvector_t *bbv = (drvector_t *)drvector_get_entry(&bbvs, i);
+        for (uint j = 0; j < bbv->entries; ++j) {
+            bb_id_count_pair_t *pair = (bb_id_count_pair_t *)drvector_get_entry(bbv, j);
+
+            const char *format_string;
+            if (j == 0 && j == bbv->entries - 1) {
+                // Only one BB and related execution count for this BBV.
+                format_string = one_pair_only;
+            } else if (j == 0) {
+                format_string = first_pair;
+            } else if (j == bbv->entries - 1) {
+                format_string = last_pair;
+            } else {
+                format_string = middle_pair;
+            }
+
+            char msg[64];
+            int len = dr_snprintf(msg, sizeof(msg) / sizeof(msg[0]), format_string,
+                                  pair->id, pair->count);
```

Comment:
style: We always use NULL_TERMINATE_BUFFER after *s*printf

Reply:
Done.